### PR TITLE
Add basic support for multiaddr addresses and improvement around peer id

### PR DESCRIPTION
**Pull request by zaibon**
relates to #65 

I'm already creating this PR so my fork doesn't diverge too far away from here.

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -4,4 +4,6 @@ pytest
 pycryptodome
 pytest-asyncio
 click
-base58
\ No newline at end of file
+base58
+pymultihash
+multiaddr
```
Comment (robzajac) [inline requirements.txt]: Please add multihash as a dependency.

File: network/swarm.py [hunk 2]
```diff
@@ -19,28 +19,24 @@ def __init__(self, my_peer_id, peerstore, upgrader):
...
+        :return: muxed connection
         """
+
         # Get peer info from peer store
         addrs = self.peerstore.addrs(peer_id)
 
         if not addrs:
             raise SwarmException("No known addresses to peer")
 
+        # TODO: define logic to choose which address to use, or try them all ?
```
Comment (alexh) [inline network/swarm.py]: I believe we should try them all and use whichever address responds the fastest

**Comment by stuckinaboot** [comment]
Hi @zaibon, is this PR intented to be merged right now? Some of the tests are failing, please fix before merging

File: network/swarm.py (hunk 2 above)
Comment (zixuanzh) [reply network/swarm.py]: This is what the go-libp2p repo is doing at the moment, and there are plans to upgrade it to a smarter dial. I forgot if it's called a traffic shaper or a dial manager. There is a bit of design freedom here. We should participate in the discussion with the rest of the community.

File: requirements.txt (hunk 1 above)
Comment (zaibon) [reply requirements.txt]: I already added the multihash lib I used, it is pymultihash, line 8

**Comment by zaibon** [comment]
@stuckinaboot 
as explained at https://github.com/zixuanzh/py-libp2p/issues/65#issuecomment-440404074
The multiaddr lib needs to be moved to the proper account and probably updated with my fixes too before we can merge this.

With the proper lib installed, I don't have any tests failing, can you point me to which test is not working with you ?

File: requirements.txt (hunk 1 above)
Comment (robzajac) [reply requirements.txt]: My mistake, thank you!

**Comment by zaibon** [comment]
What I did is locally install my fork: https://github.com/zaibon/py-multiaddr
then I develop the rest based on this multiaddr code

File: network/swarm.py (hunk 2 above)
Comment (zaibon) [reply network/swarm.py]: @zixuanzh do you have a link to the issue where they discuss this dial manager ?

**Comment by robzajac** [comment]
@zaibon I'm trying to recreate this setup - what are the steps I should follow to locally install the fork? Thanks!

**Comment by zaibon** [comment]
@robzajac 
```
cd py-libp2p
source venv/bin/activate
cd ..
git clone  https://github.com/zaibon/py-multiaddr
cd py-multiaddr
pip install .
cd ../py-libp2p
pytest -v
```

File: network/swarm.py (hunk 2 above)
Comment (zixuanzh) [reply network/swarm.py]: this is the thread https://github.com/libp2p/go-libp2p-swarm/pull/88, dial manager might be a make-up name.

**Comment by robzajac** [comment]
Hi @zaibon apologies for the pause in communication. Recently we were focusing on pushing up big design improvements over our PoC (#82 and #83).

We have reached out to @raulk regarding the discussion in #65, particular transferring ownership of sbuss/py-multiaddr to libp2p and accepting the changes in your fork. We haven't heard back yet, but I think we're ready to use your py-multiaddr fork in this repo.

Is there a way to integrate your py-multiaddr fork into our `pip` dependencies automatically, so that other contributors do not have to manually follow the steps above? Critically, we'd also like this fork to be pulled in as a dependency during CI. If we can have that set up and have this PR pass in CI I think we will be ready to merge.

Thanks for your great work!

**Comment by zaibon** [comment]
Hi @robzajac, no worries, I see you're making good progress here.

Regarding my fork. I can publish it to pypi so it can be easily integrated with pip and CI. But I guess this will be temporary until we move sbuss/py-multiaddr since I won't be able to use the same package name. But that can be done, no problem.

**Comment by zaibon** [comment]
@robzajac ok I think it's ready now. Please have another look if I didn't mess up with the code. I had to rebase a few commits.

File: host/basic_host.py [hunk 3]
```diff
@@ -56,3 +69,22 @@ def set_stream_handler(self, protocol_id, stream_handler):
         """
         stream = await self.network.new_stream(peer_id, protocol_ids)
         return stream
+
+    async def connect(self, peer_info):
```
Comment (robzajac) [inline host/basic_host.py]: Adding [reference](https://github.com/libp2p/go-libp2p/blob/master/p2p/host/basic/basic_host.go#L398) if anyone is wondering where this comes from.

**Pull request merged**